src/libopenrave-core/jsonparser/jsondownloader.cpp

```diff
@@ -0,0 +1,330 @@
+                _ParseMsgPackDocument(pContext->buffer, pContext->uri, *pContext->pDoc);
+                if (_downloadRecursively) {
+                    QueueDownloadReferenceURIs(*pContext->pDoc);
+                }
+            }
+
+            RAVELOG_DEBUG_FORMAT("successfully downloaded \"%s\", took %d[us]", pContext->uri%(currentTimestampUS-pContext->startTimestampUS));
+            ++numDownloads;
+        }
+    }
+
+    const uint64_t stopTimestampUS = utils::GetMonotonicTime();
+    RAVELOG_DEBUG_FORMAT("downloaded %d files, took %d[us]", numDownloads%(stopTimestampUS-startTimestampUS));
+}
+
+static size_t _WriteBackDataFromCurl(const char *data, size_t size, size_t dataSize, JSONDownloadContext* pContext)
+{
+    const size_t numBytes = size * dataSize;
+    std::memcpy(pContext->buffer.Push(numBytes), data, numBytes);
+    return numBytes;
+}
+
+void JSONDownloader::_QueueDownloadURI(const std::string& uri, rapidjson::Document* pDoc)
+{
+    std::string scheme, path, fragment;
+    _ParseURI(uri, scheme, path, fragment);
+
+    if (scheme.empty() && path.empty()) {
+        RAVELOG_WARN_FORMAT("unknown uri \"%s\"", uri);
+        return; // unknown uri
+    }
+    std::string canonicalUri;
+    std::string url;
+    if (scheme == "file") {
+        std::string fullFilename = RaveFindLocalFile(path);
+        if (fullFilename.empty()) {
+            RAVELOG_WARN_FORMAT("failed to resolve uri \"%s\" to local file", uri);
+            return; // no such file to download
+        }
+        url = "file://" + fullFilename;
+        canonicalUri = path;
+    }
+    else if (std::find(_vOpenRAVESchemeAliases.begin(), _vOpenRAVESchemeAliases.end(), scheme) != _vOpenRAVESchemeAliases.end()) {
+        url = _remoteUrl + path;
+        canonicalUri = scheme + ":" + path;
+    }
+    else {
+        RAVELOG_WARN_FORMAT("unable to handle uri \"%s\"", uri);
+        return; // do not understand this uri
+    }
+
+    if (!pDoc) {
+        if (_rapidJSONDocuments.find(canonicalUri) != _rapidJSONDocuments.end()) {
+            RAVELOG_VERBOSE_FORMAT("uri \"%s\" already in cache", canonicalUri);
+            return; // already in _rapidJSONDocuments
+        }
+        // create a doc and insert to map first
+        boost::shared_ptr<rapidjson::Document> pNewDoc = boost::make_shared<rapidjson::Document>(&_alloc);
+        _rapidJSONDocuments[canonicalUri] = pNewDoc;
+        pDoc = pNewDoc.get();
+    }
+
+    JSONDownloadContextPtr pContext = boost::make_shared<JSONDownloadContext>();
+    pContext->uri = canonicalUri;
+    pContext->pDoc = pDoc;
+    pContext->startTimestampUS = utils::GetMonotonicTime();
+
+    // set curl options
+    CURLcode curlCode;
+    curlCode = curl_easy_setopt(pContext->curl, CURLOPT_URL, url.c_str());
+    if (curlCode != CURLE_OK) { 
+        throw OPENRAVE_EXCEPTION_FORMAT("failed to curl_easy_setopt(CURLOPT_URL) for uri \"%s\": %s", canonicalUri%curl_easy_strerror(curlCode), ORE_InvalidArguments);
```

Comment:
These should be `ORE_CurlHandleInvalid`, the request has not been sent yet, here we are just setting up some parameters. Most likely it can only fail if there is some memory issues.